### ISSUE TITLE
Add metadata title/name sort fields

### ISSUE THREAD

**Issue by mmatyas** [mmatyas/pegasus-frontend#536]
See https://github.com/mmatyas/pegasus-frontend/issues/534#issuecomment-540003564.

**Comment by msheehan79** [comment]
@mmatyas I agree with the benefit of having a separate field to manage sort order for scenarios where the title isn't ideal, such as roman numeral sequels. I could even see this being of benefit at the collection level if alpha sorting isn't always a fit.

If you are considering expanding the games metadata, could I suggest thinking about a couple of other additional metadata fields at the game level that could be leveraged by themes? These are some custom fields I have on my metadata files that I hope to eventually leverage in a theme. Some might be a bit more niche than others but figured I'd let you know what I was thinking if you believe these might be useful elements in a theme.

Platform - If a collection spans multiple systems, it would be useful for the theme to be able to indicate which platform a given game belongs to, i.e. if it is an arcade game or its NES port.

Series - Could potentially be used by a theme to filter or group games by series.

Controller - Could be used by a theme to provide an indicator of the controller type a given game uses. Examples would be to indicate a light gun game or an arcade that uses trackball/spinner controls, or for PSX which games use the DualShock controls, or for Wii to indicate if the nunchuk or motionplus is needed etc.

One other metadata point that I currently have as a custom field on the Collections level is "Type". I use this to indicate if a given collection is a "System" collection i.e. NES Games, vs a "Custom" collection that spans across systems. I like the approach ES uses where custom collections get nested into a subfolder so it doesn't necessarily make the main collections list too long, so hopefully being able to differentiate at the collection level would allow a theme design to replicate something like this.

I know some of these might be a bit too niche, but wanted to put them out there as possible ideas to consider adding to the data model.

**Comment by iwubcode** [comment]
I think controller type would be great and not only for sorting.  A way to distinguish between mouse/keyboard and gamepad on steam would be useful but yeah other controller types would be great as well.

**Comment by mmatyas** [comment]
Sure, I'm happy to consider ideas for metadata. For adding new buildt-in fields however there is a condition I'd like to keep, that none of them should require maintaining a list of possible values inside Pegasus, which would need to get updated every time there's a new hardware release. This allows, for example, to create collections with any kind of name, without being limited to what's defined inside the frontend.

- Platforms: it's a good idea, but what I fear is that to be usable in themes it'd require maintaining a list of accepted values, which I'd prefer to avoid. Maybe there could e a separate file where one could define things like platforms and details for platforms, then the ones defined there could be used as the "platform" value. However if we'd do that, that wouldn't help theme makers too much either.

- Series: it's a good idea, I think I can add support for this.

- Controller: like platforms, this is another field where I'd need a fixed list of accepted values to be useful in themes. I could imagine a similar solution like with platforms.

In addition I was also thinking for a while now about adding support for tags -- this is similar to Steam, where games also have entries like "Single player", "Partial gamepad support" (<< @iwubcode) and such.

**Comment by mmatyas** [comment]
Also while writing this, I was thinking, if we add platform definitions, input settings, series info, etc. perhaps the metadata format could be broaden to contain everything game related, like info about the developer company, description for the series, etc., kind of like a database

**Comment by msheehan79** [comment]
@mmatyas I agree that having a static list of accepted values somewhere is going to be difficult to maintain. 

At minimum I was thinking just having the fields as free text values within Pegasus, and perhaps at the theme level if a theme was going to take advantage of any of these metadata items beyond simply rendering the value on the page they could document any expected values.

I'm not sure there is really any perfect solution to this, but I was just thinking of at least offering more options for themes. I think while it won't appeal to everyone, those who may want to offer that additional level of detail about their games could then take advantage of these extra fields.

**Comment by iwubcode** [comment]
@mmatyas  - tags would be perfect!  That might support all of this.  Games just list their specific tags and then Pegasus builds a unique list of tags as a list of filter criteria.

**Comment by mmatyas** [comment]
Ok, both tags and sort titles are now supported and documented. I've also made [a new theme](https://github.com/mmatyas/pegasus-theme-secretary) for testing purposes. Give them a try!

**Comment by iwubcode** [comment]
That theme is very interesting.  Thank you sooo much for this feature @mmatyas !  I'm sure it'll have a lot of purposes.

**Comment by msheehan79** [comment]
@mmatyas I agree the new theme is a nice way to demo the new features. I did notice however that the multi-select filters don't seem to work quite right - they only look at the first value selected, and any subsequent filters do not change the values. You have to fully empty the filtered options in order to select something different. 

Tried to figure it out myself, but I'm just not familiar enough with QML to figure out how the filter knows to refresh the list.

**Comment by mmatyas** [comment]
@msheehan79 Fixed, thanks! The filters weren't properly reevaluated on change, it should work now.

**Comment by msheehan79** [comment]
@mmatyas Yep, working now as expected. Thanks!

Any plans to add these sorts of filtering options to the default them eventually?

**Comment by mmatyas** [comment]
Yes, I'd like to improve the filter panel eventually, but I'm not sure yet when will that get done.

**Comment by msheehan79** [comment]
@mmatyas would it also be feasible to add sort title support at the collection level? 

I have a setup that is mostly system-based but also using some custom collections, and it would be nice if I could control the sort order to put all the custom collections at the end or beginning of the list, and keep the system collections grouped together without having to resort to adding an ugly prefix to each name to keep them together.

**Comment by mmatyas** [comment]
The only issue with that is `sortName` would be way too similar to `shortName`. I should find something better for `shortName`...

**Comment by msheehan79** [comment]
Here is another somewhat related scenario on the topic of sorting - in addition to the "default" sortName sorting, I was hoping to have a way to define a custom game sort order based on Collection. 

A simple example being I have 2 collections - lets say "All NES Games", and "Mario Games". The All NES collection sort would use the default alphabetical sort, but for the Mario collection I would ideally want to present a bit more of a curated list approach instead of purely alphabetical, so I would want to modify the sortNames to display games in a specific order only for that specific collection.

As I recall the game specific metadata including sortName is stored only once per game, and not per collection, so if the same game is referenced in multiple collections they all will share the same sortName.

Not sure if there is any way to incorporate a method for a collection to override the default sortName or other game metadata, or if that would end up being too complex to be worthwhile?

**Comment by mmatyas** [comment]
I see your point, but I think it might be a bit too specific to be a general field in the file format, and I also don't want to enforce any ordering on the themes. However, I could imagine this as something like an extra field: you'd add a new field for the games, eg. `extra.customOrder: 42` then in a collection select this to be used for sorting, eg. `extra.orderBy: customOrder`, and the games would be ordered by that instead of `sortTitle`.

**Comment by msheehan79** [comment]
That approach makes sense as well - I assume leveraging extra fields like this would only be possible once you add support for custom fields from the metadata, correct? Just want to make sure I'm not missing something that is already supported.

**Comment by mmatyas** [comment]
Yes, the extra fields are not currently accessible from the themes but I'd like to add support for it eventually.

**Comment by spoogemcfuck** [comment]
Is it possible to use tags to include/exclude games from collections? For example, include all games tagged "NES" and "shooter", but exclude games tagged "light-gun". This seems like a natural fit and would be really convenient, but I don't see a field for it in the metadata documentation.

If it's not already possible, would you please consider adding it?

**Comment by mmatyas** [comment]
At the moment collections filter *files*, not games, but support for some post processing based on game fields is a good idea and could be added.

**Comment by spoogemcfuck** [comment]
Excellent! Should I make a separate issue for the feature request or is this good?

**Comment by mmatyas** [comment]
Ok, I've just added #607 for it.

**Comment by mmatyas** [comment]
Both games and collections now have sort fields, we have tags support and games now know about the collections they belong to in themes. I think this issue can be closed, in case I forgot something, feel free to open new ones!
